### DISCUSSION TITLE
Package listing categories

### DISCUSSION THREAD

**Discussion by haz** [AI-Planning/planutils#127] [General]
We've officially crossed the very official threshold of having "way too many packages available": https://github.com/AI-Planning/planutils/tree/main/planutils/packages

For listing them, I think it may make sense to list top-level categories. One obvious one being `ipc2023`. Some questions...

1. Should everything be under an ontology of categories, or just group some of the big bags (like an IPC)?
2. Should we change `ipc2023` planners to reflect that it was the classical track?
3. For the author-prompted requests, do we move planners out of the `ipc2023-` category and put them top level? Might be worth doing this for the winner(s).
4. Should we start compiling & hosting the apptainer images, since they can be a beast to build? Maybe just for the non-ipc ones?

**Comment by haz** [comment]
Would specifically be good to get your thoughts on this, @FlorianPommerening , but I'd welcome anyone/everyone's input that may be interested.

**Reply by FlorianPommerening** [reply]
1. I personally don't mind the packages without a specific ontology. I'd say this is similar to how you'd see packages with `apt` (https://packages.ubuntu.com/jammy/allpackages). So if you introduce bags, I'd opt for larger ones. If it gets too fine grained, a potential problem is overlapping categories, where we don't know if a planner should go in category 1 or 2, or where we don't know which category should be nested under which.
2. I would rather introduce planners from the non-classical IPC 2023 tracks with the same naming schema (also `ipc2023-*`). I'm trying to push the idea that there is no such thing as "classical/numerical IPCs" with "optimal/satisficing/agile tracks" but that there is one IPC with a "classical optimal track", a "classical satisifcing track", ... "numerical agile track". The point would be that people can organize any set of IPC tracks independently.
3. My intuition would be to leave both: We already have that situation with some planners. For example, `cerberus-agl` is the "official version" of Cerberus (in the agile configuration) while `ipc2023_cerberus_agl` is the version that ran in the IPC 2023. I would expect `cerberus-agl` to be updated with new features if they are developed, while I would expect only bug fixes in `ipc2023_cerberus_agl`.
4. It's a bit of a hassle (where to host, who compiles them, how often should they be recompiled, who triggers that, how do we deal with CPLEX licenses). Maybe github actions together with the github container registry are an option but with the amount of time I spent building IPC images, build minutes could be a limiting factor there.

**Reply by haz** [reply]
Ok, you've convinced me on 2 and 3. For 1, I feel we at least need something to collapse the ipc2023 ones, as our list has turned into the flood you see below. For 4, I was thinking moreso up to the package maintainers. So updating the recipe to show how a dedicated github project might autobuild and host the compiled image that planutils points to.

```
Available                    Summary
---------------------------- -------------------------------------------------------------------------------------------
cerberus                     Cerberus: a red-black planning system
cerberus-agl                 Cerberus -- agile planner, post-IPC18 version
cerberus-sat                 Cerberus -- satisficing planner, post-IPC18 version
cpddl                        cpddl
cpor                         CPOR: Contingent Planning using On-line Replanning
delfi                        Delfi: online planner selection for cost-optimal planning
downward                     The Fast Downward (FD) planning system
dual-bfws-fdparser           BFWS -- FD-parser version
dual-bfws-ffparser           BFWS -- FF-parser version
enhsp                        ENHSP: Expressive Numeric Heuristic Search Planner
enhsp-2018                   ENHSP -- 2018 version
enhsp-2019                   ENHSP -- 2019 version
enhsp-2020                   ENHSP -- 2020 version
ff                           The Fast-Forward (FF) planning system
fond4ltlf                    FOND 4 LTLf
forbiditerative              The Forbid-Iterative (FI) planning system
forbiditerative-diverse-agl  FI -- diverse agile planner
forbiditerative-diverse-sat  FI -- diverse satisficing planner
forbiditerative-topk         FI -- top-k planner
forbiditerative-topq         FI -- unordered top-quality planner
hpddl2pddl                   HTN-Translation: translating HTN planning problems to PDDL
ipc2023_ApxNovelty           IPC 2023 Planner: Approximate Novelty
ipc2023_ApxNoveltyAnytime    IPC 2023 Planner: Approximate Novelty Anytime
ipc2023_ApxNoveltyTarski     IPC 2023 Planner: Grounding Schematic Representation with GRINGO for Width-based Search
ipc2023_FSM                  IPC 2023 Planner: FSM
ipc2023_NoveltyFB            IPC 2023 Planner: Forward Backward Novelty Search
ipc2023_NoveltyFBAnytime     IPC 2023 Planner: Forward Backward Anytime Novelty Search
ipc2023_SymBD_2023_opt       IPC 2023 Planner: SymBD
ipc2023_cegarplusplus        IPC 2023 Planner: CEGAR ++
ipc2023_cerberus_agl         IPC 2023 Planner: Cerberus
ipc2023_cerberus_sat         IPC 2023 Planner: Cerberus
ipc2023_complementary        IPC 2023 Planner: ComplementaryPDB
ipc2023_dalai_agl            IPC 2023 Planner: DALAI 2023 Agile
ipc2023_dalai_opt            IPC 2023 Planner: DALAI 2023 Optimal
ipc2023_dalai_sat            IPC 2023 Planner: DALAI 2023 Satisficing
ipc2023_decabstar            IPC 2023 Planner: DecAbStar
ipc2023_decstar_agl          IPC 2023 Planner: DecStar-2023
ipc2023_decstar_opt          IPC 2023 Planner: DecStar-2023
ipc2023_decstar_sat          IPC 2023 Planner: DecStar-2023
ipc2023_disco_agl            IPC 2023 Planner: DiSCO
ipc2023_disco_sat            IPC 2023 Planner: DiSCO
ipc2023_dofri                IPC 2023 Planner: Dofri
ipc2023_dom_opt              IPC 2023 Planner: QDom-Lmcut
ipc2023_fdss_2023_agl        IPC 2023 Planner: Fast Downward Stone Soup 2023 Agile
ipc2023_fdss_2023_opt        IPC 2023 Planner: Fast Downward Stone Soup 2023 Optimal
ipc2023_fdss_2023_sat        IPC 2023 Planner: Fast Downward Stone Soup 2023 Satisficing
ipc2023_fts_ff_agl           IPC 2023 Planner: FTSPlan
ipc2023_fts_ms_opt           IPC 2023 Planner: FTSPlan
ipc2023_fts_sbd_opt          IPC 2023 Planner: FTSPlan
ipc2023_hapori_delfi_agl     IPC 2023 Planner: Hapori Delfi Agile
ipc2023_hapori_delfi_opt     IPC 2023 Planner: Hapori Delfi Optimal
ipc2023_hapori_delfi_sat     IPC 2023 Planner: Hapori Delfi Satisficing
ipc2023_hapori_epsdt_agl     IPC 2023 Planner: Hapori Explainable Planner Selection - Single Decision Tree Agile
ipc2023_hapori_epsdt_opt     IPC 2023 Planner: Hapori Explainable Planner Selection - Single Decision Tree Optimal
ipc2023_hapori_epsdt_sat     IPC 2023 Planner: Hapori Explainable Planner Selection - Single Decision Tree Satisficing
ipc2023_hapori_epslr_agl     IPC 2023 Planner: Hapori Explainable Planner Selection - Linear Regression Agile
ipc2023_hapori_epslr_opt     IPC 2023 Planner: Hapori Explainable Planner Selection - Linear Regression Optimal
ipc2023_hapori_epslr_sat     IPC 2023 Planner: Hapori Explainable Planner Selection - Linear Regression Satisficing
ipc2023_hapori_greedy_agl    IPC 2023 Planner: Hapori Greedy Agile
ipc2023_hapori_greedy_opt    IPC 2023 Planner: Hapori Greedy Optimal
ipc2023_hapori_greedy_sat    IPC 2023 Planner: Hapori Greedy Satisficing
ipc2023_hapori_ibacop2_opt   IPC 2023 Planner: Hapori IBaCop2 Sat
ipc2023_hapori_ibacop2_sat   IPC 2023 Planner: Hapori IBaCop2 Sat
ipc2023_hapori_mip2_agl      IPC 2023 Planner: Hapori MIPlan All Data Random Order Agl
ipc2023_hapori_mip2_opt      IPC 2023 Planner: Hapori MIPlan Optimal All Data
ipc2023_hapori_mip2_sat      IPC 2023 Planner: Hapori MIPlan All Data Sat
ipc2023_hapori_stonesoup_agl IPC 2023 Planner: Hapori Stone Soup Agile
ipc2023_hapori_stonesoup_opt IPC 2023 Planner: Hapori Stone Soup Optimal
ipc2023_hapori_stonesoup_sat IPC 2023 Planner: Hapori Stone Soup Satisficing
ipc2023_levitron_agl         IPC 2023 Planner: Levitron Agile
ipc2023_levitron_sat         IPC 2023 Planner: Levitron Satisficing
ipc2023_maidu_agl            IPC 2023 Planner: Scorpion Maidu Agile
ipc2023_maidu_sat            IPC 2023 Planner: Scorpion Maidu Satisficing
ipc2023_odin                 IPC 2023 Planner: Odin
ipc2023_opcount4sat_agl      IPC 2023 Planner: OpCount4Sat - Agile
ipc2023_opcount4sat_sat      IPC 2023 Planner: OpCount4Sat - Satistificng
ipc2023_powerlifted_agl      IPC 2023 Planner: Powerlifted Agile
ipc2023_powerlifted_sat      IPC 2023 Planner: Powerlifted Satisficing
ipc2023_ragnarok             IPC 2023 Planner: Ragnarok
ipc2023_scorpion_2023        IPC 2023 Planner: Scorpion 2023
ipc2023_spock                IPC 2023 Planner: Spock
ipc2023_symk                 IPC 2023 Planner: Symk
ipc2023_tftm_argmax_agl      IPC 2023 Planner: TFTM1
ipc2023_tftm_argmax_sat      IPC 2023 Planner: TFTM1
ipc2023_tftm_co1_agl         IPC 2023 Planner: TFTM2
ipc2023_tftm_co1_sat         IPC 2023 Planner: TFTM2
kstar                        K* planner: integrating the K* algorithm into Fast Downward
lama                         LAMA: guiding cost-based anytime planning with landmarks
lama-first                   LAMA-first: satisficing planner without solution improvement
lapkt                        BFWS: Best-First Width Search
lifted-pddl                  Lifted PDDL: A lightweight framework for the efficient parsing and manipulation of PDDL in lifted form
lpg-td                       LPG-td: a fully automated temporal planner for PDDL2.2 domains
machetli                     Machetli: helping debugging complex programs
macq                         MAcq: The Model Acquisition Toolkit
metric-ff                    Metric Fast-Forward: extended with numerical state variables
optic                        OPTIC: Optimising Preferences and Time-Dependent Costs
planning.domains             planning.domains API tools
popf                         POPF: Partial Order Planning Forwards
poprp                        PO-PRP: Planning for Relevant Policies -- partial observability
prp                          PRP: Planning for Relevant Policies -- full observability
pyperplan                    Pyperplan: a lightweight STRIPS planner written in Python
rpmep                        RP-MEP
scorpion                     Scorpion: optimal classical planner based on saturated cost partitioning
smtplan                      SMTPlan+: a planner for hybrid systems
symk                         Sym-k: symbolic top-k planning
tarski                       Tarski: an AI planning modeling framework
tfd                          Temporal Fast Downward planning system
val                          VAL: The plan validation system
```

**Comment by FlorianPommerening** [comment]
Wouldn't it be the same list just replacing some `_`with `/`? Or what kind of hierarchy do you have in mind?

Re 4: do you mean we prepare a template for authors to use that explains how they can set up github actions etc.? Sounds good.

**Reply by haz** [reply]
> Wouldn't it be the same list just replacing some _with /? Or what kind of hierarchy do you have in mind?

Just not list them all up front. Indicate there are several under `ipc` and have `planutils list ipc` reveal them all (or, alternatively, have the `planutils list` be interactive and you select from a menu). It's just an awful lot to spit out all at once (and largely dominated by ipc planners now).

> Re 4: do you mean we prepare a template for authors to use that explains how they can set up github actions etc.? Sounds good.

Aye, precisely. We've already had some success using the existing documentation as the place to point newcomers towards.

**Reply by FlorianPommerening** [reply]
Maybe one functionality we could take from `apt` would be `apt search` This way, `planutils search ipc` would do the same as you suggested for `planutils list ipc` but without the need for hierarchy or any special treatment of any packages. It would also be relatively easy to implement in a simple way (take the input as a regex, match it against package name and description of each package, display the ones where it finds a match).

Regarding templates, I think this is a good idea but I also do not know how to use the container registry, so I'd need a template myself.

**Reply by haz** [reply]
So forgo `planutils list` entirely? Or just leave that as a not-so-advised functionality that shows the true breadth of `planutils` for those curious? :P 

Now I'm going down the rabbit hole of search indexing...exact matches could have an index built / shipped fairly easily, but anything fuzzy or semantic (e.g., equating "contest" and "competition" without having to explicitly name it) would be nice. Have any experience with this?

**Reply by FlorianPommerening** [reply]
`planutils list | grep ipc` would also be an option. I have no experience with search indexing unfortunately. For starters, I'd keep it simple and only support exact matches.
